compiler/rustc_trait_selection/src/solve/project_goals.rs

```diff
@@ -581,6 +581,19 @@ impl<'tcx> assembly::GoalKind<'tcx> for ProjectionPredicate<'tcx> {
     ) -> Vec<super::CanonicalResponse<'tcx>> {
         bug!("`Unsize` does not have an associated type: {:?}", goal);
     }
+
+    fn consider_builtin_discriminant_kind_candidate(
+        ecx: &mut EvalCtxt<'_, 'tcx>,
+        goal: Goal<'tcx, Self>,
+    ) -> QueryResult<'tcx> {
+        let discriminant = goal.predicate.self_ty().discriminant_ty(ecx.tcx());
+        let nested_goals = ecx.infcx.eq(
+            goal.param_env,
+            goal.predicate.term.ty().expect("expected ty goal"),
+            discriminant,
+        )?;
+        ecx.evaluate_all_and_make_canonical_response(nested_goals)
```

Comment:
Sorry for the churn, we can use `eq_term_and_make_canonical_response` here now. Also, this needs to happen within a probe, since all of these consider functions must have no side-effects:

```suggestion
        let discriminant = goal.predicate.self_ty().discriminant_ty(ecx.tcx());
        infcx.probe(|_| {
            ecx.eq_term_and_make_canonical_response(
                goal,
                Certainty::Yes,
                discriminant.into()
            )
        })
```

r=me after applying this though.

Reply:
don't need `discriminant.into()` i think, can be `eq_term_and_make_canonical_response(goal, Certainty::Yes, discriminant)`